runtime/src/bank/bank_hash_details.rs

```diff
@@ -0,0 +1,244 @@
+        Self {
+            version: solana_version::version!().to_string(),
+            account_data_encoding: "base64".to_string(),
+            slot,
+            bank_hash: bank_hash.to_string(),
+            parent_bank_hash: parent_bank_hash.to_string(),
+            accounts_delta_hash: accounts_delta_hash.to_string(),
+            signature_count,
+            last_blockhash: last_blockhash.to_string(),
+            accounts,
+        }
+    }
+}
+
+impl TryFrom<&Bank> for BankHashDetails {
+    type Error = String;
+
+    fn try_from(bank: &Bank) -> Result<Self, Self::Error> {
+        let slot = bank.slot();
+        if !bank.is_frozen() {
+            return Err(format!(
+                "Bank {slot} must be frozen in order to get bank hash details"
+            ));
+        }
+
+        // This bank is frozen; as a result, we know that the state has been
+        // hashed which means the delta hash is Some(). So, .unwrap() is safe
+        let accounts_delta_hash = bank
+            .rc
+            .accounts
+            .accounts_db
+            .get_accounts_delta_hash(slot)
+            .unwrap()
+            .0;
+        let mut accounts = bank
+            .rc
+            .accounts
+            .accounts_db
+            .get_pubkey_hash_account_for_slot(slot);
+        // get_pubkey_hash_account_for_slot() returns an arbitrary ordering;
+        // sort by pubkey to match the ordering used for accounts delta hash
+        accounts.sort_by_key(|(pubkey, _, _)| *pubkey);
+
+        Ok(Self::new(
+            slot,
+            bank.hash(),
+            bank.parent_hash(),
+            accounts_delta_hash,
+            bank.signature_count(),
+            bank.last_blockhash(),
+            BankHashAccounts(accounts),
+        ))
+    }
+}
+
+// Wrap the Vec<...> so we can implement custom Serialize/Deserialize traits on the wrapper type
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct BankHashAccounts(pub Vec<(Pubkey, Hash, AccountSharedData)>);
+
+#[derive(Deserialize, Serialize)]
+/// Used as an intermediate for serializing and deserializing account fields
+/// into a human readable format.
+struct TempAccount {
+    pubkey: String,
+    hash: String,
+    owner: String,
+    lamports: u64,
+    rent_epoch: Epoch,
+    executable: bool,
+    data: String,
+}
+
+impl Serialize for BankHashAccounts {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.0.len()))?;
+        for (pubkey, hash, account) in self.0.iter() {
+            let temp = TempAccount {
+                pubkey: pubkey.to_string(),
+                hash: hash.to_string(),
+                owner: account.owner().to_string(),
+                lamports: account.lamports(),
+                rent_epoch: account.rent_epoch(),
+                executable: account.executable(),
+                data: BASE64_STANDARD.encode(account.data()),
```

Comment:
could isolate this with a `impl From<&AccountSharedData> for TempAccount`

Reply:
Not quite, the account pubkey and hash are not within `AccountSharedData`. Hence, the tuple:
```rust
 Vec<(Pubkey, Hash, AccountSharedData)>
 ```
 Hypothetically, I could replace the tuple with a struct with named fields like:
 ```rust
 pub struct AccountDetails {
     pubkey: Pubkey,
     hash: Hash,
     data: AccountSharedData
}
```
and do `impl From<&AccountDetails> for SerdeAccount` and vice versa (I renamed `TempAccount` to `SerdeAccount`). Maybe late night, apathetic brain speaking, but I don't see doing the above as a huge win.

Reply:
Done (in both directions)